### PR TITLE
gh-105566: Deprecate unusual ways of creating `typing.NamedTuple` classes

### LINKED ISSUE #105566: Deprecate unusual ways of creating `typing.NamedTuple` classes
# Feature or enhancement

I propose that we deprecate in Python 3.13 the following unusual ways of constructing a `typing.NamedTuple`:

```py
from typing import NamedTuple

Foo = NamedTuple("Foo", x=int, y=int)  # NamedTuple with "x" and "y" fields
Bar = NamedTuple("Bar")                # empty NamedTuple
Baz = NamedTuple("Baz", None)          # empty NamedTuple
Eggs = NamedTuple("Eggs", fields=None) # empty NamedTuple
```

# Pitch

`typing.NamedTuple` has been around for quite a while now, but none of the above methods of constructing `NamedTuple`s are supported by type checkers. If they're still unsupported by type checkers after all this time, they're unlikely to ever be supported by type checkers.

Deprecating, and eventually removing, these ways of constructing `NamedTuple`s will allow us to simplify the code at runtime. It will also be less confusing for users. Every way in which the runtime and type checkers differ in behaviour is a potential point of confusion for users; in general, we should work to keep these points of difference to a minimum.

These methods of constructing `NamedTuple`s are not commonly seen in the wild. They're also pretty redundant -- if you want to construct a `NamedTuple` in one line, in a single function call, you can do it like this, which is supported by type checkers:

```py
Foo = NamedTuple("Foo", [("x", int), ("y", int)])
```

If you want to construct an empty `NamedTuple`, meanwhile, you can do it in one of the following two ways, which are both supported by type checkers:

```py
Bar = NamedTuple("Bar", [])
class Baz(NamedTuple): ...
```

# Previous discussion

For very similar reasons, we previously deprecated and removed the keyword-argument syntax for creating TypedDicts. This was deprecated in 3.11, and removed in 3.13:
- #90224
- #104786

### Linked PRs
* gh-105609

### PR DESCRIPTION
**Pull request by AlexWaygood**
Closes #105566

* Issue: gh-105566

----
:books: Documentation preview :books:: https://cpython-previews--105609.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -2038,6 +2038,19 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      (``NT = NamedTuple("NT", x=int)``) is deprecated, and will be removed in
+      3.15.
+
+   .. deprecated-removed:: 3.13 3.15
+      When using the functional syntax to create a NamedTuple class, failing to
+      pass a value to the 'fields' parameter (``NT = NamedTuple("NT")``) is
+      deprecated. Passing ``None`` to the 'fields' parameter
+      (``NT = NamedTuple("NT", None)``) is also deprecated. Both will be
+      removed in Python 3.15. To create a NamedTuple class with 0 fields, use
+      ``class NT(NamedTuple): ...`` or ``NT = NamedTuple("NT", [])``.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
      ``class NT(NamedTuple): pass`` or ``NT = NamedTuple("NT", [])``.
```
Didn't you tell me off recently for using `...`? :)

I think `pass` is more idiomatic for an intentionally empty class.

File: Lib/typing.py [hunk 2]
```diff
@@ -2753,7 +2753,16 @@ def __new__(cls, typename, bases, ns):
...
 
 
-def NamedTuple(typename, fields=None, /, **kwargs):
+class _Sentinel:
+    __slots__ = ()
+    def __repr__(self):
+        return '<sentinel>'
+
+
+_sentinel = _Sentinel()
```
Comment (sobolevn) [inline Lib/typing.py:2762]: I think that `= object()` is good enough for this case.

**Review by sobolevn — COMMENTED** [review]
Good cleanup!

File: Lib/typing.py (hunk 2 above)
Comment (AlexWaygood) [reply Lib/typing.py:2762]: I disagree. Using `= object()` would mean that the displayed signature would be very ugly if somebody did `help(typing.NamedTuple)` in the REPL

Comment (AlexWaygood) [reply Lib/typing.py:2762]: Also, we'll be able to reuse the `_Sentinel` class when we do #105570 :)

File: Doc/library/typing.rst (hunk 1 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst]: > Didn't you tell me off recently for using `...`? :)

Heh, very true :)

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -2038,6 +2038,19 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      (``NT = NamedTuple("NT", x=int)``) is deprecated, and will be disallowed
+      in 3.15. Use the class-based syntax or the functional syntax instead.
+
+   .. deprecated-removed:: 3.13 3.15
+      When using the functional syntax to create a NamedTuple class, failing to
+      pass a value to the 'fields' parameter (``NT = NamedTuple("NT")``) is
+      deprecated. Passing ``None`` to the 'fields' parameter
+      (``NT = NamedTuple("NT", None)``) is also deprecated. Both will be
+      disallowed in Python 3.15. To create a NamedTuple class with 0 fields,
+      use ``class NT(NamedTuple): ...`` or ``NT = NamedTuple("NT", [])``.
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      use ``class NT(NamedTuple): pass`` or ``NT = NamedTuple("NT", [])``.
```

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -133,6 +133,17 @@ Deprecated
...
+* Creating a :class:`typing.NamedTuple` class using keyword arguments to denote
+  the fields (``NT = NamedTuple("NT", x=int, y=int)``) is deprecated, and will
+  be disallowed in Python 3.15. Use the class-based syntax or the functional
+  syntax instead. (Contributed by Alex Waygood in :gh:`105566`.)
+* When using the functional syntax to create a :class:`typing.NamedTuple`
+  class, failing to pass a value to the 'fields' parameter
+  (``NT = NamedTuple("NT")``) is deprecated. Passing ``None`` to the 'fields'
+  parameter (``NT = NamedTuple("NT", None)``) is also deprecated. Both will be
+  disallowed in Python 3.15. To create a NamedTuple class with 0 fields, use
+  ``class NT(NamedTuple): ...`` or ``NT = NamedTuple("NT", [])``.
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  ``class NT(NamedTuple): pass`` or ``NT = NamedTuple("NT", [])``.
```

File: Misc/NEWS.d/next/Library/2023-06-09-20-34-23.gh-issue-105566.YxlGg1.rst [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Deprecate creating a :class:`typing.NamedTuple` class using keyword
+arguments to denote the fields (``NT = NamedTuple("NT", x=int, y=str)``).
+This will be disallowed in Python 3.15.
+Use the class-based syntax or the functional syntax instead.
+
+Two methods of creating ``NamedTuple`` classes with 0 fields using the
+functional syntax are also deprecated, and will be disallowed in Python 3.15:
+``NT = NamedTuple("NT")`` and ``NT = NamedTuple("NT", None)``. To create a
+``NamedTuple`` class with 0 fields, either use ``class NT(NamedTuple): ...`` or
```
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-06-09-20-34-23.gh-issue-105566.YxlGg1.rst]:
```suggestion
``NamedTuple`` class with 0 fields, either use ``class NT(NamedTuple): pass`` or
```

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**